Initial new lwpa_poll API using context structs

samkearney opened the pull request:
This contains only the Windows changes for now.

samkearney committed:
Initial new lwpa_poll API using context structs

samkearney committed:
Documentation and cleanup - new poll API

samkearney committed:
Better 'platform-defined' syntax for Doxygen

samkearney committed:
new-poll-api: Update Changelog

samkearney committed:
Add user data passing functionality to socket poll API

nwagner commented on src/lwpa/socket.dox:
```diff
@@ -231,23 +239,97 @@ lwpa_error_t lwpa_socket(unsigned int family, unsigned int type, lwpa_socket_t *
+ *
+ *  \param[in,out] context Pointer to LwpaPollContext from which to remove socket.
+ *  \param[in] socket Socket to remove.
+ *  \return #kLwpaErrOk: Socket removed successfully.\n
```
I'm not sure it's necessary to have this function return anything.  Is it really an error if the context was bad or the socket wasn't there?

samkearney replied:
Yeah, good point. I'm torn on the pattern of returning status from deinit() or remove() functions. I agree that I don't see the usefulness, yet I see it in a lot of other code. If you think it's silly, I will make sure we're consistent about making all of our deinit() and remove() functions void.

nwagner replied:
I like consistency above all. 

 If there's a case where a real system error could occur that the user should react to, then we need to return something.  Otherwise, we shouldn't. And maybe in the former case, we don't return those errors, or in the docs stress that they are warnings and not failures.

samkearney replied:
I think if you are getting an error from a deinit function, it is probably an indication that there is some bug in your use of the functions. In this case the error code _might_ be useful for debug logging. But you'd probably be debugging anyway. I dunno.

samkearney replied:
Alright, I'll harmonize on void deinit and remove functions.

nwagner commented on src/lwpa/socket.dox:
```diff
@@ -231,23 +239,97 @@ lwpa_error_t lwpa_socket(unsigned int family, unsigned int type, lwpa_socket_t *
+ *  resources being used to track polling context.
+ *
+ *  \param[in] context Pointer to LwpaPollContext to deinitialize.
+ *  \return #kLwpaErrOk: Deinit successful.\n
```
Is it necessary to return anything here?

nwagner commented on src/lwpa_plat/windows/lwpa/socket.c:
```diff
@@ -669,45 +881,82 @@ int lwpa_poll(LwpaPollfd *fds, size_t nfds, int timeout_ms)
+      if (LWPA_FD_ISSET(sock_desc->socket, writefds))
       {
-        fd->revents |= LWPA_POLLOUT;
+        if (sock_desc->events & LWPA_POLL_CONNECT)
```
I'm a bit confused by LWPA_POLL_CONNECT.  When does that event get cleared?  Otherwise the user will never see LWPA_POLL_OUT, right?

samkearney replied:
LWPA_POLL_CONNECT is an abstraction around the fact that different operating systems require different polling methods for nonblocking connects. For example, on Windows using select(), you need to add a connecting socket to both writefds and exceptfds to handle errors. On other platforms I believe it's only writefds.

So this flag is necessary to differentiate nonblocking connects from general polling for writability (i.e. a write to a full TCP pipe returns EWOULDBLOCK, and the app wants to poll for when there is buffer space again).

samkearney replied:
More to your point, theoretically once the socket is connected, the user will modify the socket so that it is not interested in the LWPA_POLL_CONNECT event anymore. Then they can utilize LWPA_POLL_OUT if desired.

samkearney replied:
I'd be open to the idea of making LWPA_POLL_CONNECT an auto-clearing event...

nwagner replied:
How do they modify the socket?  Do they have to remove & re-add it?

samkearney replied:
> How do the modify the socket? Do they have to remove & re-add it?

Right now yes, I believe that's how they'd have to do it.

nwagner replied:
It feels cleaner to me to have a function to clear the CONNECT flag.  I could be persuaded to make this auto-clearing, too, since you presumably want to write once you've connected..

samkearney replied:
As far as the function to clear goes, how about just a generic modify_socket() function where you can change the set of events however you want?

nwagner replied:
I originally wasn't sure if you wanted to add such a general feature for the specific CONNECT issue, but that I think that's probably the right way to do it.

nwagner commented on src/lwpa_plat/windows/lwpa/socket.c:
```diff
@@ -669,45 +881,82 @@ int lwpa_poll(LwpaPollfd *fds, size_t nfds, int timeout_ms)
+      if (LWPA_FD_ISSET(sock_desc->socket, exceptfds))
       {
-        fd->revents |= LWPA_POLLPRI;
+        if (sock_desc->events & LWPA_POLL_CONNECT)
```
I'm a bit confused by LWPA_POLL_CONNECT.  When does that event get cleared?  Otherwise the user will never see LWPA_POLL_OOB, right?

samkearney committed:
Implement changes from code review:
- Add lwpa_poll_modify_socket() function to change events on a socket
- Make deinit() and remove_socket() functions void

samkearney committed:
lwpa_poll: Add meaningful error code for the case of waiting on a context with no sockets

samkearney committed:
poll: update documentation
